File: GameEngine_IOS/GameEngine/Commons/WaveFront/OBJLoader.m
#import "OBJLoader.h"

/**
 * Parses files formated in wavefront format
 */
@implementation OBJLoader
{
@private
    
    /**
     *	Positions where the coordinates are when each line of wavefront
     *is parsed
     */
    enum
    {
        posX = 1,
        posY,
        posZ
    } CoordinatePositions;
    
    enum {
        vertexPos = 0,
        texturPos,
        normalPos
    };
    
}

/**
 * Vertices positions in the model
 */
const NSString *VERTEX_PREFIX = @"v";

/**
 * Normal vertices
 */
const NSString *NORMAL_PREFIX = @"vn";

/**
 * Vertices texture coordinates in our model
 */
const NSString *TEXTURE_PREFIX = @"vt";

/**
 * Face each face represents one triangle
 */
const NSString *FACE_PREFIX = @"f";

/**
 * Extension of the obj files
 */
NSString* OBJ_EXTENSION = @"obj";

/**
 *  Max size of a line in the object file
 */
const int MAX_LINE_LENGTH = 512;

const int INITIAL_ARRAY_CAPACITY = 1000;

/**
 * Number of components depending upon the type of element
 */
const int COORDINATES_BY_VERTEX = 3;
const int COORDINATES_BY_TEXTURE = 2;
const int COORDINATES_BY_NORMAL = 3;

/**
 * Parse two strings and return the equivalent vector2f
 *
 * @param xStr
 *            X component
 * @param yStr
 *            Y component
 *
 * @return The parsed vector
 */
+ (Vector2f*) parseVector2f : (NSString*) xStr : (NSString*) yStr {
    float x = [xStr floatValue];
    float y = [yStr floatValue];
    
    return [[Vector2f alloc] init: x : y];
}


/**
 * Parse three strings and return the equivalent vector3f
 *
 * @param xStr
 *            X component
 * @param yStr
 *            Y component
 * @param zStr
 *            Z component
 *
 * @return The parsed vector
 */
+ (Vector3f*) parseVector3f :(NSString*) xStr : (NSString*) yStr :  (NSString*) zStr {
    float x = [xStr floatValue];
    float y = [yStr floatValue];
    float z = [zStr floatValue];
    
    return [[Vector3f alloc] init: x : y : z];
}

/**
 * Uses the list of elements of the waveFront file to create one shape
 *
 * @param vertices	List of vertices from the waveFront shape
 * @param normals	List of normals from the waveFront shape
 * @param textures	List of texture coordinates from the waveFront shape
 * @param facesLst 	List of face from the waveFront shape
 
 * @return The waveFront element as an IShape
 */
+ (id<IShape>) createShape : (NSMutableArray<Vector3f*>*) vertices :
(NSMutableArray<Vector3f*>*) normals :
(NSMutableArray<Vector2f*>*) textures :
(NSMutableArray<PolygonalFace*>*) facesLst {
    // This are the format of data accepted by the loader
    // We setup the arrays now that we know the size of them
    int numberVertices = [vertices count];
    int numberFaces = [facesLst count];
    float verticesArray[numberVertices * COORDINATES_BY_VERTEX];
    float normalsArray[numberVertices * COORDINATES_BY_NORMAL];
    float texturesArray[numberVertices * COORDINATES_BY_TEXTURE];
    unsigned short indicesArray[numberFaces];
    
    for (int j = 0; j < numberFaces; j++) {
        PolygonalFace* face = [facesLst objectAtIndex:j];
        
        int vertexIndex = [face vertexIndex];
        int normalIndex = [face normalIndex];
        int textureIndex = [face textureIndex];
        
        // Point to loaded structure

        Vector3f* vertice = [vertices objectAtIndex: vertexIndex];
        Vector3f* currentNorm = [normals objectAtIndex: normalIndex];
        Vector2f* currentTexture = [textures objectAtIndex: textureIndex];
        
        // Build index lists
        indicesArray[j] = vertexIndex;
        
        // Uses the (faces and vertices list to build the final vertices
        // array
        verticesArray[vertexIndex * COORDINATES_BY_VERTEX] = vertice.x;
        verticesArray[vertexIndex * COORDINATES_BY_VERTEX + 1] = vertice.y;
        verticesArray[vertexIndex * COORDINATES_BY_VERTEX + 2] = vertice.z;
        
        
        // Build the normals list
        normalsArray[vertexIndex * COORDINATES_BY_NORMAL] = currentNorm.x;
        normalsArray[vertexIndex * COORDINATES_BY_NORMAL + 1] = currentNorm.y;
        normalsArray[vertexIndex * COORDINATES_BY_NORMAL + 2] = currentNorm.z;
        
        // Build the texture coordinates list
        if (currentTexture != nil) {
            texturesArray[vertexIndex * COORDINATES_BY_TEXTURE] = currentTexture.x;
            texturesArray[vertexIndex * COORDINATES_BY_TEXTURE + 1] = currentTexture.y;
        }
    }
    
    
    return [[WfObject alloc] initWithAVertices:verticesArray
                                     aCountVertices: (numberVertices * COORDINATES_BY_VERTEX)
                                aTextureCoordinates: texturesArray
                           aCountTextureCoordinates:(numberVertices * COORDINATES_BY_TEXTURE)
                                           aNormals: normalsArray
                                      aCountNormals: (numberVertices * COORDINATES_BY_NORMAL)
                                           aIndices: indicesArray
                                      aCountIndices: numberFaces];
    

}

/**
 * Parses one waveFront file
 *
 * @param objFileName
 *            Name of the file without extension
 *
 * @return The shape with information about the waveFront file read
 */
+ (id<IShape>) loadObjModel : (NSString*) objFileName {
    
    //Declares the arrays where the elements of the obj file are going to be store
    NSMutableArray<Vector3f*> *vertices = [[NSMutableArray alloc]initWithCapacity: INITIAL_ARRAY_CAPACITY];
    NSMutableArray<Vector2f*> *textures =  [[NSMutableArray alloc]initWithCapacity: INITIAL_ARRAY_CAPACITY];
    NSMutableArray<Vector3f*> *normals =  [[NSMutableArray alloc]initWithCapacity: INITIAL_ARRAY_CAPACITY];
    NSMutableArray<PolygonalFace*> *facesLst =  [[NSMutableArray alloc]initWithCapacity: INITIAL_ARRAY_CAPACITY];
    
    //Open the obj file from the disk
    NSString* objPath = [[NSBundle mainBundle] pathForResource:objFileName ofType: OBJ_EXTENSION];
    FILE *file = fopen([objPath UTF8String], "r");
    char buffer[MAX_LINE_LENGTH];
    //Read the obj file line by line
    while ((fgets(buffer, sizeof(char) * MAX_LINE_LENGTH, file) != NULL)){
        NSString* line = [NSString stringWithUTF8String:buffer];
        NSArray *currentLine = [line componentsSeparatedByString:@" "];
        //Switch of the type element to read
        NSString *prefix = [currentLine objectAtIndex: 0];
        typedef void (^CaseElementTypeBlock)();
        
        NSDictionary *d = @{
                            VERTEX_PREFIX:
                                ^{
                                    // Parses the vertices
                                    Vector3f *vertex = [self parseVector3f: [currentLine objectAtIndex: posX] :
                                                        [currentLine objectAtIndex: posY] : [currentLine objectAtIndex: posZ]];
                                    [vertices addObject: vertex];
                                },
                            TEXTURE_PREFIX:
                                ^{
                                    // Parses the texture coordinates
                                    Vector2f *texture = [self parseVector2f: [currentLine objectAtIndex: posX] :
                                                         [currentLine objectAtIndex: posY]];
                                    [textures addObject: texture];
                                },
                            NORMAL_PREFIX:
                                ^{
                                    // Parses the normals
                                    Vector3f *normal = [self parseVector3f: [currentLine objectAtIndex: posX] :
                                                        [currentLine objectAtIndex: posY] : [currentLine objectAtIndex: posZ]];
                                    [normals addObject: normal];
                                },
                            FACE_PREFIX:
                                ^{
                                    // Parses the faces
                                    for (int i = 1; i < 4; i++) {
                                        NSArray *fVertexStr = [[currentLine objectAtIndex: i] componentsSeparatedByString:@"/"];
                                        PolygonalFace *pFace = [self processVertex : fVertexStr];
                                        
                                        [facesLst addObject:pFace];
                                    }
                                }
                            };
        if([d objectForKey: prefix] == nil) {
            //Default block
        } else {
            ((CaseElementTypeBlock)d[[currentLine objectAtIndex: 0]])(); // invoke the correct block of code
        }
    }
    fclose(file);
    
    return [self createShape : vertices : normals : textures : facesLst];
}

/**
 * Process one vertex
 */
+ (PolygonalFace*) processVertex : (NSArray*) vertexData {
    unsigned short vertexIndex = [[vertexData objectAtIndex: vertexPos] integerValue] - 1;
    unsigned short textureIndex = [[vertexData objectAtIndex: texturPos] integerValue] - 1;
    unsigned short normalIndex = [[vertexData objectAtIndex: normalPos] integerValue] - 1;
    
    
    return [[PolygonalFace alloc] init : vertexIndex : textureIndex : normalIndex];
}

@end

File: GameEngine_IOS/GameEngine/Commons/WaveFront/OBJLoader.swift
import Foundation

/**
 * Parses files formated in wavefront format
 */
public class OBJLoader : GenericLoader {
    
    /**
     * Extension of the obj files
     */
    private static let OBJ_EXTENSION : String = "obj";
    
    
    private static let INITIAL_ARRAY_CAPACITY : Int = 1000;
    
    /* Character for split material and group */
    private static let MAT_GROUP_SPLIT : String = "@";
    
    /**
     * Number of components depending upon the type of element
     */
    private static let COORDINATES_BY_VERTEX : Int = 3;
    private static let COORDINATES_BY_TEXTURE : Int = 2;
    private static let COORDINATES_BY_NORMAL : Int = 3;
    
    /**
     *
     * @param face
     *            reference to a Polygonal face
     *
     * @return A key to use in the dictionary of indices
     */
    private static func getKey(face : PolygonalFace) -> String {
        let materialName :  String = (face.materialName == EMPTY_STRING) ? EMPTY_STRING : face.materialName;
        let groupName : String = (face.groupName == EMPTY_STRING) ? EMPTY_STRING : face.groupName;
        return materialName + MAT_GROUP_SPLIT + groupName;
    }
    
    /**
     * Uses the hash map to create a list of objects
     *
     * @param verticesArray
     *            Array of vertices
     * @param normalsArray
     *            Array of normals
     * @param texturesArray
     *            Array of textures positions
     * @param indicesArrayMap
     *            Dictionary with indices that make the faces
     * @param materials
     *            HashMap of materials that make part of the object (if any)
     *
     * @return List of objects created
     */
    private static func buildShapesLst(verticesArray : [Float], _ normalsArray : [Float], _ texturesArray : [Float],
                                       _ indicesArrayMap : Dictionary<String, [Int]>, _ materials : Dictionary<String, IExternalMaterial>?) -> [IShape] {
        var wfObjectList : [IShape] = [IShape]();
        
        for (key, indicesList) in indicesArrayMap {
            let materialName = key.componentsSeparatedByString(MAT_GROUP_SPLIT)[0];
            let groupName = key.componentsSeparatedByString(MAT_GROUP_SPLIT)[1];
            var material : IExternalMaterial? = nil;
            
            if((materials != nil) && (materials![materialName] != nil)){
                material = materials![materialName];
            }
            let shape : IShape = WfObject(verticesArray, texturesArray, normalsArray, indicesList, groupName, material);
            wfObjectList.append(shape)
        }
        return wfObjectList;
    }
    
    /**
     * Uses the list of elements of the waveFront file to create one shape
     *
     * @param vertices	List of vertices from the waveFront shape
     * @param normals	List of normals from the waveFront shape
     * @param textures	List of texture coordinates from the waveFront shape
     * @param faces 	List of face from the waveFront shape
     * @param materials HashMap of materials that make part of the object (if any)
     
     * @return The waveFront element as an IShape
     */
    private static func createShapes(vertices : Array<Vector3f>,
                                     _ normals : Array<Vector3f>,
                                       _ textures : Array<Vector2f>,
                                         _ faces : Array<PolygonalFace>,
                                           _ materials : Dictionary<String, IExternalMaterial>?) -> [IShape] {
        
        // This are the format of data accepted by the loader
        // We setup the arrays now that we know the size of them
        let numberVertices = vertices.count;
        
        //Vertices
        let verticesLength = (numberVertices * COORDINATES_BY_VERTEX);
        var aVertices : Array<Float> = Array<Float>(count: verticesLength, repeatedValue: 0.0);
        
        //Normals
        let normalsLength = numberVertices * COORDINATES_BY_NORMAL;
        var normalsArray : Array<Float> = Array<Float>(count: normalsLength, repeatedValue: 0.0);
        
        //Textures
        let textureLength = numberVertices * COORDINATES_BY_TEXTURE;
        var texturesArray : Array<Float> = Array<Float>(count: textureLength, repeatedValue: 0.0);
        
        //Indices
        var indicesArrayMap : Dictionary<String, [Int]> = Dictionary<String, [Int]>();
        
        for face in faces {
            
            let vertexIndex = face.vertexIndex;
            let normalIndex = face.normalIndex;
            let textureIndex = face.textureIndex;
            
            // Point to loaded structure
            
            let  vertice : Vector3f = vertices[vertexIndex];
            let  currentNorm : Vector3f = normals[normalIndex];
            let  currentTexture : Vector2f! = textures[textureIndex];
            
            // Build index lists
            let indicesKey = getKey(face)
            
            if(indicesArrayMap[indicesKey] == nil)
            {
                var indicesArray : [Int] = [Int]();
                indicesArray.append(vertexIndex);
                indicesArrayMap[indicesKey] = indicesArray;
            } else
            {
                var indicesArray : [Int] = indicesArrayMap[indicesKey]!;
                indicesArray.append(vertexIndex);
                indicesArrayMap[indicesKey] = indicesArray
            }
            
            
            // Uses the (faces and vertices list to build the final vertices array
            
            aVertices[vertexIndex * COORDINATES_BY_VERTEX + 0] = vertice.x;
            aVertices[vertexIndex * COORDINATES_BY_VERTEX + 1] = vertice.y;
            aVertices[vertexIndex * COORDINATES_BY_VERTEX + 2] = vertice.z;
            
            // Build the normals list
            normalsArray[vertexIndex * COORDINATES_BY_NORMAL + 0] = currentNorm.x;
            normalsArray[vertexIndex * COORDINATES_BY_NORMAL + 1] = currentNorm.y;
            normalsArray[vertexIndex * COORDINATES_BY_NORMAL + 2] = currentNorm.z;
            
            
            
            // Build the texture coordinates list
            if (currentTexture != nil) {
                texturesArray [vertexIndex * COORDINATES_BY_TEXTURE + 0] = currentTexture.x;
                texturesArray [vertexIndex * COORDINATES_BY_TEXTURE + 1] = 1.0 - currentTexture.y;
            }
        }
        
        return buildShapesLst(aVertices, normalsArray, texturesArray, indicesArrayMap, materials);
    }
    
    /**
     * Process one vertex
     */
    private static func processVertex(vertexData : Array<String>, _ groupName : String, _ materialName : String) -> PolygonalFace {
        let vertexIndex : Int = (Int(vertexData[TComponentPosition.vertexPos.rawValue]) ?? 0) - 1;
        let textureIndex : Int = (Int(vertexData[TComponentPosition.texturPos.rawValue]) ?? 0) - 1;
        let normalIndex : Int = (Int(vertexData[TComponentPosition.normalPos.rawValue]) ?? 0) - 1;
        
        
        return PolygonalFace(vertexIndex: vertexIndex,
                             textureIndex: textureIndex,
                             normalIndex: normalIndex,
                             groupName: groupName,
                             materialName: materialName
        );
    }
    
    /**
     * Parses one waveFront file
     *
     * @param objFileName
     *            Name of the file without extension
     *
     * @return The shape with information about the waveFront file read
     */
    public static func loadObjModel(objFileName : String) -> [IShape] {
        //Declares the arrays where the elements of the obj file are going to be store
        var materials : Dictionary<String, IExternalMaterial>?  = nil;
        var vertices : Array<Vector3f> = Array<Vector3f>();
        var textures : Array<Vector2f> = Array<Vector2f>();
        var normals : Array<Vector3f> = Array<Vector3f>();
        var faces : Array<PolygonalFace> = Array<PolygonalFace>();
        var currentGroupName : String = EMPTY_STRING;
        var currentMaterialName : String = EMPTY_STRING;
        
        
        //Open the obj file from the disk
        let objPath : String = NSBundle.mainBundle().pathForResource(objFileName, ofType: OBJ_EXTENSION)!
        let file : UnsafeMutablePointer<FILE> = fopen(objPath, "r")
        
        if(file != nil) {
            
            let buffer : UnsafeMutablePointer<Int8> = UnsafeMutablePointer<Int8>.alloc(MAX_LINE_LENGTH);
            //Read the obj file line by line
            let bytesToRead : Int32 = Int32(sizeof(CChar) * MAX_LINE_LENGTH);
            while(fgets(buffer, bytesToRead, file) != nil) {
                let linen = String(UTF8String: UnsafePointer<CChar>(buffer))
                let line = linen?.stringByReplacingOccurrencesOfString("\n", withString: EMPTY_STRING);
                let currentLine : Array<String> = line!.componentsSeparatedByString(GenericLoader.SPLIT_TOKEN);
                let prefix : String = currentLine[0];
                switch(prefix) {
                // The comments in the OBJ do nothing
                case OBJPrefix.COMMENT:
                    break;
                // Definition of the list of materials of the model
                case OBJPrefix.MATERIALS:
                    let materialsFileName = parseStringComponent(line!, currentLine)
                    materials = MtlLoader.loadObjModel(materialsFileName);
                    break;
                    // Parses the information that is to update the current material
                // used
                case OBJPrefix.USE_MATERIAL:
                    currentMaterialName = parseStringComponent(line!, currentLine);
                    break;
                /// Define the name of the object or the group
                case OBJPrefix.OBJECT_NAME:
                    currentGroupName = parseStringComponent(line!, currentLine);
                    break;
                case OBJPrefix.GROUP:
                    currentGroupName = parseStringComponent(line!, currentLine);
                    break;
                /// Defines the smooth shading (Not done yet)
                case OBJPrefix.SMOOTH_SHADING:
                    break;
                // Parses the vertices
                case OBJPrefix.VERTEX:
                    // Parses the vertices
                    let vertex : Vector3f = self.parseVector3f(currentLine[TCoordinatePosition.posX.rawValue], yStr: currentLine[TCoordinatePosition.posY.rawValue], zStr: currentLine[TCoordinatePosition.posZ.rawValue]);
                    vertices.append(vertex);
                    break;
                // Parses the texture coordinates
                case OBJPrefix.TEXTURE:
                    // Parses the texture coordinates
                    let texture : Vector2f = self.parseVector2f(currentLine[TCoordinatePosition.posX.rawValue], yStr: currentLine[TCoordinatePosition.posY.rawValue]);
                    textures.append(texture);
                    break;
                // Parses the normals
                case OBJPrefix.NORMAL:
                    // Parses the normals
                    let normal : Vector3f = self.parseVector3f(currentLine[TCoordinatePosition.posX.rawValue], yStr: currentLine[TCoordinatePosition.posY.rawValue], zStr: currentLine[TCoordinatePosition.posZ.rawValue]);
                    normals.append(normal);
                    break;
                // Parses the faces
                case OBJPrefix.FACE:
                    // Parses the faces
                    for i : Int in 1 ..< 4 {
                        let fVertexStr : Array<String> = currentLine[i].componentsSeparatedByString("/");
                        let pFace : PolygonalFace = self.processVertex(fVertexStr, currentGroupName, currentMaterialName);
                        faces.append(pFace);
                    }
                    break;
                default:
                    print("Impossible to parse:\(line)")
                    break
                }
            }
            
            free(buffer);
            fclose(file);
        }
        
        return self.createShapes(vertices, normals, textures, faces, materials);
        
    }
    
}
